xorg: allow overriding via overrideScope'

thefloweringash opened the pull request:
###### Motivation for this change
Wanted to modify some xorg.* packages in an overlay, ran into https://github.com/NixOS/nixpkgs/issues/11944. I did a fairly straight forward conversion into `makeScope` / `overrideScope`, but I don't have a good idea of the state of the art in this area.

Additionally discovered that while `pkgs/servers/x11/xorg/default.nix` is auto-generated, it has been manually updated for a few packages. I updated the source lists so that regenerating `default.nix` doesn't change any packages. The exception to this is `xeyes`, which gained a new buildInput. The source lists should be converted to https, but I'd prefer to do that in a separate PR.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

thefloweringash committed:
xorg: allow overriding via overrideScope

thefloweringash committed:
xorg: sync source lists with default.nix

thefloweringash commented on pkgs/servers/x11/xorg/default.nix:
```diff
@@ -1,2771 +1,2994 @@
     builder = ./builder.sh;
     src = fetchurl {
       url = http://xcb.freedesktop.org/dist/libxcb-1.12.tar.bz2;
       sha256 = 
```
Moved this to `overrides.nix`. It got lost in the regeneration.

thefloweringash replied:
I believe this is the only rebuild in this PR.

thefloweringash commented on pkgs/servers/x11/xorg/overrides.nix:
```diff
@@ -98,243 +99,247 @@ in
       sed "/^Requires:/s/$/, freetype2/" -i "$dev/lib/pkgconfig/xft.pc"
     '';
-  };
+    passthru = {
```
The generated file no longer applies `// { inherit $buildInputs; }` to each package. Qt3 relies on being able to access `libXft.freetype` and `libXft.fontconfig`. Updated to `passthru`.

thefloweringash commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13464,7 +13464,7 @@ with pkgs;
     abiCompat = config.xorg.abiCompat # `config` because we have no `xorg.override`
       or (if stdenv.isDarwin then "1.18" else null); # 1.19 needs fixing on Darwin
-  } // { inherit xlibsWrapper; } );
+  } // { inherit xlibsWrapper; } ) [ "packages" ]);
```
Need to remove the `"packages"` attribute from `xorg`'s scope. Otherwise it gets merged into the main attribute set and makes `lambdabot` fail. I don't yet know why it's not a problem to have `"callPackage"`, etc from `makeScope` 

https://github.com/NixOS/nixpkgs/blob/8014c7f186afb66ae746d6b1af9ce4510fb55eae/pkgs/top-level/splice.nix#L27-L28

https://github.com/NixOS/nixpkgs/blob/8014c7f186afb66ae746d6b1af9ce4510fb55eae/lib/customisation.nix#L194-L203

They should probably also be removed. Should this be done in `all-packages.nix` or `splice.nix`?

thefloweringash reviewed (commented):
The generator script attempts to use a caching feature of `nix-prefetch-url` which no longer exists. Every time it's run it downloads everything from scratch.

thefloweringash committed:
xorg: add caching to generate-expr-from-tarballs.pl

thefloweringash committed:
xorg: simplify argument passing into xorg and overrides

thefloweringash committed:
xorg: don't splice scope attributes into main package set

Ericson2314 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13467,21 +13467,16 @@ with pkgs;
-    inherit (buildPackages) pkgconfig xmlto asciidoc flex bison;
-    inherit (darwin) apple_sdk cf-private libobjc;
+  xorg = recurseIntoAttrs ((lib.callPackagesWith pkgs ../servers/x11/xorg {
+  }).overrideScope (lib.callPackageWith pkgs ../servers/x11/xorg/overrides.nix {
```
Can we do inline https://github.com/NixOS/nixpkgs/blob/f305b9b0862f28a85a3dcc2a517d4fcb345abc42/lib/customisation.nix#L201 to get something like
```
(lib.fixedPoints.extends (import ../servers/x11/xorg/overrides.nix) (import ../servers/x11/xorg)
```
(This ignores the flip. That would be for the odd order of `self: super:`, when you already did idiomatic `self: super:`.)

Ericson2314 commented:
> Why did the old version import some dependencies for xorg via `inherit (buildPlatform) ...;`? Does this break cross?

So first of all, I think you meant `buildPackages`. And to answer your question, yes, it is needed because `callPackagesWith pkgs` is being used directly. See the unfortunate "splicing' stuff that goes into normal `callPcakges`. `callPackagesWith pkgs` doesn't get that, and so is missing packages.

Now that `splicePackages` is exposed, however, you can let-bind a `callPackageWithoutXorg` which will obviate the need to sort specific packages by hand.

Ericson2314 commented:
Overall, great work! It would be nice to get this into 18.09 as it's a good cleanup and will make any xorg backports easier if we do so.

thefloweringash commented:
Rebased to resolve the conflict, fixed `callPackagesWith` vs `callPackageWith`.

It sounds like you've got a clear idea of how to make this nicer, but I'm not familiar enough with the splicing to grasp it. Can you expand a bit on how implementing `callPackageWithoutXorg` would look?

As for using`lib.fixedPoints.extends`, would that require changing `default.nix` and `overrides.nix` to take `pkgs` explicitly instead of being called with `callPackage`?

Thanks for looking at this!

Ericson2314 commented:
I'm going to make a few fixes here, then merge soon.

Ericson2314 reviewed (changes requested):
N.B. to other reviews I will fix cross

thefloweringash commented:
I'm very confused by this. I don't see any merge conflicts when doing a merge locally. I'm guessing whatever is causing the merge conflicts is also causing the mass rebuild tags. Assuming ofborg builds the GitHub automatic merge, then it will be building this old version:

```
$ git fetch origin pull/45328/merge && git log -1 FETCH_HEAD
commit c03453c1270eb1405b6d5687b74e284c2c185834
Merge: f799dff21bd 1b748ec40ea
Author: Andrew Childs <lorne@cons.org.nz>
Date:   Sun Sep 23 12:10:31 2018 +0000

    Merge 1b748ec40ea89b344dc20fae33fda3f3dc28897c into f799dff21bd70738a0724419752dbb8d9ed3ce6b
```

For reference:
- current master: 9179c1449db7b868504551cff9972acf6c1ce4f9
- current PR: 0e88495e3eb8a602b58c50331995e1af0b96ef4f
- current PR without dummy commit: b663eb04e5146bc7e64fa9613a3706393e9ae3e2

Ericson2314 committed:
xorg: Don't use deprecated item

Ericson2314 committed:
Merge remote-tracking branch 'upstream/master' into xorg-override

Ericson2314 commented:
@dtzWill Sorry just saw. Sure that makes sense, will do!
